src/qibo/backends/numpy.py

```diff
@@ -479,55 +507,64 @@ def execute_circuit_repeated(self, circuit, initial_state=None, nshots=None):
                             gate.substitute_symbols()
                         state = gate.apply(self, state, nqubits)
 
+            if circuit.density_matrix:
+                final_states.append(state)
             if circuit.measurements:
-                result = CircuitResult(self, circuit, state, 1)
+                result = CircuitResult(state, circuit.measurements, self, 1)
                 sample = result.samples()[0]
                 results.append(sample)
                 if not circuit.density_matrix:
-                    probabilities += result.probabilities()
+                    probabilities += result.probabilities(target_qubits)
                     samples.append("".join([str(s) for s in sample]))
-            else:
-                results.append(state)
 
-        if circuit.measurements:
-            final_result = CircuitResult(self, circuit, state, nshots)
-            final_result._samples = self.aggregate_shots(results)
-            if not circuit.density_matrix:
-                final_result._repeated_execution_probabilities = probabilities / nshots
-                final_result._repeated_execution_frequencies = (
-                    self.calculate_frequencies(samples)
+        if circuit.density_matrix:  # this implies also it has_collapse
+            assert circuit.has_collapse
+            final_state = np.asarray(final_states).mean(0)
+            if circuit.measurements:
+                qubits = [q for m in circuit.measurements for q in m.target_qubits]
+                final_result = CircuitResult(
+                    final_state, circuit.measurements, self, nshots
                 )
+                probabilities = final_result.probabilities(qubits)
+                final_result._repeated_execution_probabilities = probabilities
+                circuit._final_state = final_result
+                return final_result
+            else:
+                final_result = QuantumState(final_state, self)
+                probabilities = final_result.probabilities()
+                final_result._repeated_execution_probabilities = probabilities
```

Comment:
Regarding the `QuantumState._repeated_execution_probabilities` (comment above), I think this is the only place they are used. However if I follow the code properly I think it is just doing a trivial caching. You are calculating the probabilities from the `final_result` and then you are setting them back to the same object. If you call `final_result` again after that you will get the values cached in `_repeated_execution_probabilities` but these will be the same with just redoing the calculation. If that's correct, it should be safe to drop this variable completely.

If you need caching, you can handle it internally in the `QuantumState` object, or use [standard Python methods.](https://docs.python.org/3/library/functools.html#functools.cache). However I don't think that's needed in our case.